### PR TITLE
Add pre-commit config with pre-commit-hooks

### LINKED ISSUE #1593: Add pre-commit configuration file
To prevent the annoyance of realizing I forgot to run black after I have pushed a commit, I think it would be good to add a pre-commit hook configuration file to automatically run black, flake8, and other formatting packages.

**Are you willing to help implement and maintain this feature?** Yes

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

Add [pre-commit](https://github.com/pre-commit/pre-commit) hooks to enforce certain rules on all files in the repo.

The following rules are applied:
- check-added-large-files
- check-yaml
- end-of-file-fixer
- trailing-whitespace
- forbid-crlf
- remove-crlf
- chmod

Note that `pre-commit run --all-files` has been added to the Makefile, so running `make format` locally or `/format` on a GitHub PR will apply the formatting rules.

TODO:
- [x] Add a `.pre-commit-config.yaml` file for standard pre-commit hooks
- ~Configure [pre-commit-ci](https://pre-commit.ci) GitHub App~ not using, doing `pre-commit` manually ourselves

References:
- https://pre-commit.com/index.html#quick-start

Fixes #1593

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash command is:

- `/format`: automatically format and lint the code

### CODE REVIEW

**Comment by weiji14** [comment]
Results from first pre-commit.ci run at https://results.pre-commit.ci/run/github/85352251/1717479459.EAFPxChRS_aFLgAcskCvfQ:

```
check for added large files..............................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing doc/_templates/autosummary/class.rst
Fixing doc/_templates/autosummary/exception.rst
Fixing pygmt/tests/data/contours.txt
Fixing doc/_templates/autosummary/function.rst
Fixing doc/team.md
Fixing CODE_OF_CONDUCT.md

trim trailing whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing examples/projections/README.txt
Fixing doc/changes.md
```

Do we want to try applying the autofixes directly in this PR?

**Comment by seisman** [comment]
Yes to make sure that autofix works as expected.

**Comment by weiji14** [comment]
pre-commit.ci autofix

File: .pre-commit-config.yaml
Comment (weiji14) [inline .pre-commit-config.yaml:10]: Can also add these hooks from https://github.com/Lucas-C/pre-commit-hooks that handle LF file endings and 644 permissions?

```suggestion
    - id: trailing-whitespace
- repo: https://github.com/Lucas-C/pre-commit-hooks
  rev: v1.5.5
  hooks:
    - id: forbid-crlf
    - id: remove-crlf
    - id: chmod
      args: ['644']
```

Comment (seisman) [reply .pre-commit-config.yaml:10]: Looks good.

**Comment by seisman** [comment]
Don't forget to add the `.pre-commit-config.yaml` file to `https://github.com/GenericMappingTools/pygmt/blob/main/MANIFEST.in`.

File: .pre-commit-config.yaml
Comment (weiji14) [reply .pre-commit-config.yaml:10]: Hmm, not sure why, but pre-commit.ci seeems to think our files have 664 permission instead of 644?

Comment (weiji14) [reply .pre-commit-config.yaml:10]: I've added back the GitHub Action workflows for 644 permission checks/formatting for now at commit 5a01ad61b354a832fd956deb8da90b22ac1e22e1.

File: doc/contributing.md
Comment (weiji14) [inline doc/contributing.md:489]: I'm thinking of whether to just add `pre-commit run --all-files` to the Makefile, or have it listed separately. Main consideration is that Windows users won't have `make` installed by default, so they would only be able to run `pre-commit run --all-files`.

File: README.md
Comment (seisman) [inline README.md:11]: I feel this badge is not necessary.

File: doc/contributing.md
Comment (seisman) [reply doc/contributing.md:489]: `make` is added as a development dependency in `environment.yml`, so even Windows developers have `make` installed.

**Comment by seisman** [comment]
Also need to mention `pre-commit.ci autofix` in the PR template.

File: doc/contributing.md
Comment (weiji14) [reply doc/contributing.md:489]: Oh yes, forgot that it is possible to conda install make now! I'll add it to the Makefile then.

File: README.md
Comment (weiji14) [reply README.md:11]: Would it be more useful if it links to https://results.pre-commit.ci/repo/github/85352251 instead?

**Review by weiji14 — COMMENTED** [review]
> Also need to mention `pre-commit.ci autofix` in the PR template.

Done at b3eee6ff6a4d77bf0af24f30d71b5331c8faf526. But I'm also wondering if this is redundant, because I've added `pre-commit run --all-files` to the `make format` command, so typing `/format` will also run pre-commit now :slightly_smiling_face:

**Comment by seisman** [comment]
> But I'm also wondering if this is redundant, because I've added `pre-commit run --all-files` to the `make format` command, so typing `/format` will also run pre-commit now

Adding `pre-commit run --all-files` to `make format` means that users/maintainers must set up `pre-commit` locally (although they just need to run `pre-commit install` once).

What about adding `pre-commit run --all-files` to https://github.com/GenericMappingTools/pygmt/blob/main/.github/workflows/format-command.yml, so that `pre-commit` is run when `/format` is executed?

File: README.md
Comment (seisman) [reply README.md:11]: pre-commit CI is unlikely to fail after this PR is merged. So I guess the badge and the status report is not necessary.

**Comment by weiji14** [comment]
> > But I'm also wondering if this is redundant, because I've added `pre-commit run --all-files` to the `make format` command, so typing `/format` will also run pre-commit now
> 
> Adding `pre-commit run --all-files` to `make format` means that users/maintainers must set up `pre-commit` locally (although they just need to run `pre-commit install` once).

As long as they have `pre-commit` installed, they don't need to run `pre-commit install` because `make format` and `pre-commit run --all-files` will automatically set up pre-commit on the first run. E.g.:

```bash
$ pre-commit clean  # uninstall/clean out pre-commit files
$ make format
ruff check --fix --exit-zero pygmt doc/conf.py examples
All checks passed!
ruff format pygmt doc/conf.py examples
301 files left unchanged
pre-commit run --all-files
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/Lucas-C/pre-commit-hooks.
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/Lucas-C/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
check for added large files..............................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
CRLF end-lines checker...................................................Passed
CRLF end-lines remover...................................................Passed
```

> What about adding `pre-commit run --all-files` to https://github.com/GenericMappingTools/pygmt/blob/main/.github/workflows/format-command.yml, so that `pre-commit` is run when `/format` is executed?

Depends on whether we want `pre-commit` in the Makefile or not. I'm actually thinking if we want to install `ruff` as a pre-commit hook (in a separate PR), and then `make format` would just be an alias for `pre-commit run --all-files`.

File: README.md
Comment (weiji14) [reply README.md:11]: Technically, `pre-commit run` would only run on changed files by default (unless you use the `--all files` flag), so there is a chance that the `pre-commit` CI check will pass on PRs, but fail on the main branch if some lint rules have changed on other files. In that case, it might be good to have this badge to see what the error is (or we can click on the red cross next to the commit msg if logged into GitHub).

**Comment by seisman** [comment]
> As long as they have `pre-commit` installed, they don't need to run `pre-commit install` because `make format` and `pre-commit run --all-files` will automatically set up pre-commit on the first run.

Good to know. Then we don't need the pre-commit.ci service anymore, right?

**Comment by weiji14** [comment]
> > As long as they have `pre-commit` installed, they don't need to run `pre-commit install` because `make format` and `pre-commit run --all-files` will automatically set up pre-commit on the first run.
> 
> Good to know. Then we don't need the pre-commit.ci service anymore, right?

Yes, pre-commit.ci isn't absolutely necessary since we can run it via `style_checks.yml`. Only other reason to use `pre-commit.ci` is to have the quarterly updates for the pre-commit hook versions. It doesn't seem like dependabot supports updating .pre-commit-config.yaml yet according to https://github.com/dependabot/dependabot-core/issues/1524.

**Comment by seisman** [comment]
I'm OK with either keeping pre-commit.ci or not.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (seisman) [inline .github/PULL_REQUEST_TEMPLATE.md:21]: Changes in this file can be reverted.

File: .github/workflows/style_checks.yaml
Comment (seisman) [inline .github/workflows/style_checks.yaml:41]:
```suggestion
        run: |
           make check
           pre-commit run --all-files
```

File: .github/workflows/style_checks.yaml
Comment (seisman) [inline .github/workflows/style_checks.yaml:37]:
```suggestion
          python -m pip install ruff pre-commit
```

File: .github/workflows/format-command.yml
Comment (seisman) [inline .github/workflows/format-command.yml:35]:
```suggestion
          python -m pip install ruff pre-commit
```

File: README.md
Comment (seisman) [reply README.md:11]: The CI status badge can be removed after we run `pre-commit` in the "Style Checks" workflow.

File: doc/contributing.md
Comment (seisman) [inline doc/contributing.md:488]: Changes in this file can be reverted.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (weiji14) [reply .github/PULL_REQUEST_TEMPLATE.md]: Done at 386ad8e9eba7f4a8ebfac4a13805db91d6b89c46

File: .github/workflows/style_checks.yaml
Comment (weiji14) [reply .github/workflows/style_checks.yaml]: Done at 25652fc6b52298921f76914d003c17637287c067

File: .github/workflows/style_checks.yaml
Comment (weiji14) [reply .github/workflows/style_checks.yaml]: Done at 25652fc6b52298921f76914d003c17637287c067

File: .github/workflows/format-command.yml
Comment (weiji14) [reply .github/workflows/format-command.yml]: Done at 25652fc6b52298921f76914d003c17637287c067

File: README.md
Comment (weiji14) [reply README.md]: Removed at d937cacf0d43ba76749ef6b6476a75b38bfd685d

File: doc/contributing.md
Comment (weiji14) [reply doc/contributing.md]: Done at 0f09fb605e4a4248c0ac0ca4766fb91299ddb6aa

**Comment by weiji14** [comment]
Ok, I've removed the `pre-commit.ci` GitHub App installation from the repo. Let me try the 644 permissions check again to see if it works on GitHub Actions.

**Review by seisman — APPROVED** [review]
Looks great.

**Pull request merged**
